src/components/application_manager/src/application_impl.cc

```diff
@@ -511,13 +511,24 @@ void ApplicationImpl::SuspendStreaming(
     application_manager_.OnAppStreaming(app_id(), service_type, false);
     sync_primitives::AutoLock lock(video_streaming_suspended_lock_);
     video_streaming_suspended_ = true;
+
+    if (video_streaming_approved()) {
```

Comment:
I don't see any guarantee that another app is streaming when this app is disallowed, so it seems to me like this would just result in `OnVideoDataStreaming(false)` not being sent even when no app is streaming. This would result in the wrong state for the HMI. A better check would be to check if another app is streaming to determine whether to send this message or not.

Reply:
Thank you for your review.

OnVideoDataStreaming(false) will be sent before this app loses approval.

Application loses its approval at ApplicationImpl::StopNaviStreaming(). This function will be called from ApplicationImpl::StopStreaming or ApplicationImpl::StopStreamingForce. Both function call ApplicationImpl::SuspendStreaming() (this sends OnVideoDataStreaming(false) to HMI)
before calling ApplicationImpl::StopNaviStreaming().

So I think OnVideoDataStreaming(false) is already sent when this if statement become false.

Reply:
Ah, I believe that I may have been confusing this with `video_streaming_allowed()`, which is modified outside of this class.

Is there any reason that this statement is not put above the rest of the statement, as is done in the `WakeUpStreaming` method?

```cpp
  if (video_streaming_approved()) {
    LOG4CXX_INFO(
        logger_,
        "Do not suspend video streaming because already not approved");
    return;
  }
  video_stream_suspend_timer_.Stop();
  application_manager_.OnAppStreaming(app_id(), service_type, false);
  sync_primitives::AutoLock lock(video_streaming_suspended_lock_);
  video_streaming_suspended_ = true;
  MessageHelper::SendOnDataStreaming(
      service_type, false, application_manager_);
```

Reply:
My understanding was that you were saying that the issue was that `SuspendStreaming` was being called twice for the same service, which would mean that `video_streaming_suspended_` was already set to true if `video_streaming_approved()` is false. 

> So I think OnVideoDataStreaming(false) is already sent when this if statement become false.

If this double-call is not the cause of this issue, I can't see what could be, as these two functions are the only places where the `SendOnDataStreaming` method is used.

Reply:
Consider following scenario:
There are 2 video streaming apps.
1. App1 becomes FULL and starts navi and audio services and become approved.
2. App1 will send onVideoStreaming(true)
3. User switches App2 to FULL.
4. App1 becomes BACKGROUND and stops streaming
	4-1. App1 sends onVideoStreaming(false)
	4-2. App1 becomes unapproved
5. App2 starts streaming
	5-1. App2 becomes approved
	5-2. App2 sends onVideoStreaming(true)

If there is some potential timing issue that step 5-2 is executed before step 4-1, App2's streaming will be stopped right after streaming started. In this situation, App1 should not send onVideoStreaming(false) but should clear flags. This is my implementation.

In any case, this discussion is for abnormal scenerio. If you have a strong reason for early return is preferrable, I will change PR.

Reply:
These flags are guaranteed to have been cleared if `video_streaming_approved` is false. As you have pointed out, `video_streaming_approved` is never set to false until after `SuspendStreaming` is called.

Your scenario just seems to bring us back to my first comment, `video_streaming_approved` tells us nothing about whether another app is streaming or not.

Using your scenario:
With these changes, if 5-2 happens before 4-1 the HMI will see
1. onVideoStreaming(true)
2. onVideoStreaming(false)

Which is the exact scenario this change is meant to avoid. This if statement does not prevent this, because `video_streaming_approved` is still true when `SuspendStreaming` is called for app 1, since 4-2 hasn't happened yet.

Reply:
Thanks for checking. I understand it.

I updated PR. Thanks!

Reply:
I'm sorry I made you misunderstand.

I do not have a reproducing steps about the scenario which I explained (5-2 happens before 4-1).
When I created this PR, I wondered which to choose how to implement SuspendStreaming(). When app do not have approval, app should do "early return (* 1)" or "do not send onVideoDataStreaming (false) but clear other flags (*2)".

As my first decision, I thought that the latter (*2) would be better for the case where the problem described in the previous comment "potentialy" exist.
However, since you commented that above scenario will not occur, I changed it to (*1). 

Therefore, my last commit does not fix specific issue. I just changed the implementation style.